fpmsyncd/fpmsyncd.cpp

```diff
@@ -75,11 +118,30 @@ int main(int argc, char **argv)
              */
             pipeline.flush();
 
-            cout << "Waiting for fpm-client connection..." << endl;
-            fpm.accept();
-            cout << "Connected!" << endl;
+            shared_ptr<Selectable> link;
+            if (useFpmLink)
+            {
+                shared_ptr<FpmLink> fpm = make_shared<FpmLink>(&sync);
+
+                cout << "Waiting for fpm-client connection..." << endl;
+                fpm->accept();
+                cout << "Connected!" << endl;
 
-            s.addSelectable(&fpm);
+                link = fpm;
+            }
+            else
+            {
+                shared_ptr<NetLink> netlink = make_shared<NetLink>();
+
+                netlink->registerGroup(RTNLGRP_IPV4_ROUTE);
+                netlink->registerGroup(RTNLGRP_IPV6_ROUTE);
+                netlink->registerGroup(RTNLGRP_MPLS_ROUTE);
+                cout << "NetLink listening for route messages..." << endl;
+                netlink->dumpRequest(RTM_GETROUTE);
+
+                link = netlink;
+            }
```

Comment:
I'm trying understand this code. Is this used when there is no fast path (ASIC forwarding), i.e. standalone linux. Doesn't zebra already have a netlink interface to program routes in linux IP stack.

Reply:
FRR does not forward MPLS-related information to FPM.  This support is missing from FRR Zebra.  This was discussed at length in a previous MPLS sub-community meeting.  So if you program routes or NHs that have MPLS components using either iproute2 or FRR vtysh, fpmsyncd will not receive this information on FpmLink.  Using either Juniper cRPD or Metaswitch as routing stack, this is not an issue, but neither of these routing stacks are open-source.  For unit-testing purposes I had switched fpmsyncd to listen directly to the NetLink socket rather than FPM socket, since the Linux kernel correctly encodes MPLS into the NetLink message.  At the MPLS sub-community meeting, it was agreed that I would contribute this capability of switching fpmsyncd so that unit-tests could be run by the general community.

Reply:
Can you point the unit test using this code. If it's not being used then I suggest we remove it until there is a need for it.

Reply:
All the unit-tests for MPLS use this code.  It is not possible to push MPLS Routes and NHs using iproute2 without it.  See MPLS test cases in sonic-swss/tests/test_route.py in this PR.

Reply:
I dont think it is needed as part of MPLS support. You can have it as a separate PR and unblock this PR.

Reply:
This content has been moved to a separate PR (https://github.com/Azure/sonic-swss/pull/1765)